hal/armv7a/pmap.c

```diff
@@ -201,22 +203,20 @@ addr_t pmap_destroy(pmap_t *pmap, int *i)
 
 void _pmap_switch(pmap_t *pmap)
 {
-	if (pmap->asid_ix == 0)
+	if (pmap->asid == ASID_INVALID) {
 		_pmap_asidAlloc(pmap);
-	else if (hal_cpuGetTTBR0() == (pmap->addr | TTBR_CACHE_CONF))
+	}
+	else if (hal_cpuGetTTBR0() == ((u32)pmap->addr | TTBR_CACHE_CONF)) {
 		return;
+	}
 
 	/* Assign new user's page dir to TTBR0 register */
-	hal_cpuDataSyncBarrier();
 	hal_cpuSetContextId(0);
 	hal_cpuInstrBarrier();
 	hal_cpuSetTTBR0(pmap->addr | TTBR_CACHE_CONF);
 	hal_cpuInstrBarrier();
-	hal_cpuSetContextId((u32)pmap->pdir | pmap_common.asids[pmap->asid_ix]);
-
-	/* TODO: invalidate TLB only if asids pool is run out.
-       This code should be moved to _pmap_asidAlloc and _pmap_asidDealloc */
-	hal_cpuInvalTLB();
+	hal_cpuSetContextId(((u32)pmap->pdir & 0xff) | pmap->asid);
```

Comment:
That doesn't seem right, `pdir`'s valid bits begin with bit 12th, this way you pretty much made `pdir` irrelevant

Reply:
Thanks for pointing this out!